Trustroots/trustroots
# Refactor OfferLocation React map

Refactor `OfferLocation` component and create a new base component for all the maps.

**Biggest changes;**
- I think the API of the new library is nicer and more capable, which will be important for refactoring the main search feature to React.
- I think this component structure makes more sense; to create a map you now just need to import our base `Map` component and it takes care of all the basic layers, components, etc. Then you'll add your "content" to the map either as HTML or by creating new `BaseLayer`s:
    ```jsx
    <Map zoom={5} className="my-map">
        <div>
            My content
            ...could be e.g. BaseLayer providing map markers or simple <SVGOverlay />.
        </div>
    </Map>
    ```
    ...you can see this in action in `<OfferLocation>` and `<OfferLocationOverlay>` components.
- Vector tiles 🎉 — much faster and smoother experience compared to bitmap tiles 

![Screen Recording 2020-03-15 at 16 28 23 2](https://user-images.githubusercontent.com/87168/76703450-76c8db80-66da-11ea-845a-9ba443c69734.gif)


#### Proposed Changes

* Create a new base `Map` component and use a different library (([`react-map-gl`](https://www.npmjs.com/package/react-map-gl) instead of [`react-leaflet`](https://www.npmjs.com/package/react-leaflet)).
  * Has fallback capability to render OSM tiles when Mapbox API key isn't provided, so we don't need to be vendor-locked because of this.
  * Use vector tiles instead of bitmap tiles. Vectors look visually better while zooming and it's also possible to rotate (or tilt) the map, although I've disabled that feature for now.
  * Adds [Mapbox attribution](https://docs.mapbox.com/help/how-mapbox-works/attribution/) which we apparently should've had previously.
* Refactor `OfferLocation` component to use above base component (used at profiles' offer section and at "meet" list)
* Adds scale control to OfferLocation map

#### Testing Instructions

1. **Hosting map**
    * Create a hosting offer
    * Open your profile
    * Map should show up under the "Accommodation" section
    <img width="354" alt="Screenshot 2020-03-15 at 15 46 54" src="https://user-images.githubusercontent.com/87168/76702924-4d0db580-66d6-11ea-8a7f-121661f5823b.png">
    <img width="412" alt="Screenshot 2020-03-15 at 15 44 26" src="https://user-images.githubusercontent.com/87168/76702930-54cd5a00-66d6-11ea-8114-f992547d0d1e.png">

2. **Meet map**
    * Create a meet offer
    * Go to "meet" section and see the map rendered

    <img width="925" alt="Screenshot 2020-03-15 at 15 44 35" src="https://user-images.githubusercontent.com/87168/76702939-63b40c80-66d6-11ea-8eb2-7b3fbd6911d9.png">


3. **Additionally...**
    - Try different screen widths, "yes" and "maybe" hosting offer, try zooming in and out by scrolling, clicking controls, double-clicking...
    - Compared to the maps at the live site
    - When changing screen size (E.g. change browser's width), the map doesn't refresh to new dimensions. I'd like to leave fixing that to a follow-up PR.
